File: tools/hygiene/sweep.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import subprocess
from dataclasses import dataclass, asdict
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
DEFAULT_PATHS = [ROOT / "tmp", ROOT / "logs", ROOT / "cortical-loop" / "logs"]
DEFAULT_MIGRATIONS_DIR = ROOT / "migrations"

SEVERITY_WEIGHT = {"info": 1, "warn": 4, "critical": 10}


@dataclass
class Finding:
    check: str
    severity: str  # info|warn|critical
    title: str
    message: str
    metadata: dict[str, Any]
    recoverable: bool = False
    cleaned: bool = False


@dataclass
class SweepSummary:
    mode: str
    safe: bool
    dry_run: bool
    risk_score: int
    counts: dict[str, int]
    cleaned_items: int
    findings: list[Finding]


class HygieneSweep:
    def __init__(
        self,
        stale_session_minutes: int,
        stale_file_days: int,
        oversized_log_mb: int,
        paths: list[Path],
        migrations_dir: Path,
        db: str,
        verbose: bool = False,
    ):
        self.stale_session_minutes = stale_session_minutes
        self.stale_file_days = stale_file_days
        self.oversized_log_bytes = oversized_log_mb * 1024 * 1024
        self.paths = paths
        self.migrations_dir = migrations_dir
        self.db = db
        self.verbose = verbose

    def run(self, mode: str, safe: bool, dry_run: bool) -> SweepSummary:
        findings: list[Finding] = []
        findings.extend(self._check_subagent_sessions())
        findings.extend(self._check_stale_files(mode=mode, safe=safe, dry_run=dry_run))
        findings.extend(self._check_duplicate_migration_prefixes())
        findings.extend(self._check_oversized_session_logs(mode=mode, safe=safe, dry_run=dry_run))

        counts = {"info": 0, "warn": 0, "critical": 0}
        cleaned_items = 0
        weighted = 0
        for f in findings:
            counts[f.severity] = counts.get(f.severity, 0) + 1
            weighted += SEVERITY_WEIGHT.get(f.severity, 1)
            if f.cleaned:
                cleaned_items += 1

        max_possible = max(1, len(findings) * SEVERITY_WEIGHT["critical"])
        risk_score = min(100, round((weighted / max_possible) * 100))

        return SweepSummary(
            mode=mode,
            safe=safe,
            dry_run=dry_run,
            risk_score=risk_score,
            counts=counts,
            cleaned_items=cleaned_items,
            findings=findings,
        )

    def _check_subagent_sessions(self) -> list[Finding]:
        findings: list[Finding] = []
        stale_ms = self.stale_session_minutes * 60 * 1000

        data = self._read_subagent_runtime_data()
        if data is None:
            # Fallback: infer stale/orphaned work from task queue state.
            stale_minutes = int(self.stale_session_minutes)
            stale_in_progress = self._psql_json(
                f"""
                SELECT COALESCE(json_agg(t), '[]'::json)
                FROM (
                  SELECT id, title, status, assigned_to, created_at,
                         EXTRACT(EPOCH FROM (NOW() - created_at)) * 1000 AS runtime_ms
                  FROM cortana_tasks
                  WHERE status='in_progress'
                    AND created_at < NOW() - INTERVAL '{stale_minutes} minutes'
                  ORDER BY created_at ASC
                  LIMIT 25
                ) t;
                """
            )
            stale_in_progress = stale_in_progress if isinstance(stale_in_progress, list) else []

            if stale_in_progress:
                findings.append(
                    Finding(
                        check="subagent_sessions",
                        severity="warn",
                        title="Potentially orphaned in-progress tasks",
                        message=f"Detected {len(stale_in_progress)} long-running in_progress task(s) beyond stale threshold.",
                        metadata={"threshold_minutes": stale_minutes, "tasks": stale_in_progress},
                    )
                )
            else:
                findings.append(
                    Finding(
                        check="subagent_sessions",
                        severity="info",
                        title="No stale in-progress tasks",
                        message="Fallback queue scan found no likely orphaned/stale tasks.",
                        metadata={"threshold_minutes": stale_minutes},
                    )
                )
            return findings

        active = data.get("active") or []
        recent = data.get("recent") or []

        stale_active = [s for s in active if int(s.get("runtimeMs") or 0) > stale_ms]
        if stale_active:
            findings.append(
                Finding(
                    check="subagent_sessions",
                    severity="critical",
                    title="Stale active subagent sessions",
                    message=f"{len(stale_active)} running subagent session(s) exceeded stale threshold.",
                    metadata={
                        "threshold_minutes": self.stale_session_minutes,
                        "sessions": [
                            {
                                "runId": s.get("runId"),
                                "label": s.get("label"),
                                "runtimeMs": s.get("runtimeMs"),
                                "status": s.get("status"),
                            }
                            for s in stale_active
                        ],
                    },
                )
            )
        else:
            findings.append(
                Finding(
                    check="subagent_sessions",
                    severity="info",
                    title="No stale active subagent sessions",
                    message="All active subagent sessions are within freshness threshold.",
                    metadata={"active_count": len(active), "threshold_minutes": self.stale_session_minutes},
                )
            )

        problematic_recent = [s for s in recent if str(s.get("status")) in {"failed", "timeout"}]
        if problematic_recent:
            findings.append(
                Finding(
                    check="subagent_sessions",
                    severity="warn",
                    title="Recent failed/timed-out subagent sessions",
                    message=f"Detected {len(problematic_recent)} recent subagent failures/timeouts.",
                    metadata={
                        "sessions": [
                            {
                                "runId": s.get("runId"),
                                "label": s.get("label"),
                                "status": s.get("status"),
                                "runtimeMs": s.get("runtimeMs"),
                            }
                            for s in problematic_recent
                        ]
                    },
                )
            )
        return findings

    def _read_subagent_runtime_data(self) -> dict[str, Any] | None:
        try:
            proc = subprocess.run(
                ["openclaw", "subagents", "list", "--json"],
                capture_output=True,
                text=True,
                cwd=str(ROOT),
            )
            if proc.returncode != 0:
                return None
            data = json.loads(proc.stdout)
            if isinstance(data, dict):
                return data
        except Exception:  # noqa: BLE001
            return None
        return None

    def _psql_json(self, sql: str) -> Any:
        env = os.environ.copy()
        env["PATH"] = f"/opt/homebrew/opt/postgresql@17/bin:{env.get('PATH', '')}"
        proc = subprocess.run(
            ["psql", self.db, "-v", "ON_ERROR_STOP=1", "-At", "-c", sql],
            text=True,
            capture_output=True,
            env=env,
        )
        if proc.returncode != 0:
            return None
        raw = (proc.stdout or "").strip()
        if not raw:
            return None
        try:
            return json.loads(raw)
        except json.JSONDecodeError:
            return None

    def _all_candidate_files(self) -> list[Path]:
        files: list[Path] = []
        for base in self.paths:
            if not base.exists() or not base.is_dir():
                continue
            for p in base.rglob("*"):
                if p.is_file():
                    files.append(p)
        return files

    def _check_stale_files(self, mode: str, safe: bool, dry_run: bool) -> list[Finding]:
        findings: list[Finding] = []
        now = datetime.now(timezone.utc).timestamp()
        stale_seconds = self.stale_file_days * 86400

        stale_files: list[Path] = []
        for p in self._all_candidate_files():
            try:
                age = now - p.stat().st_mtime
            except FileNotFoundError:
                continue
            if age >= stale_seconds:
                stale_files.append(p)

        if not stale_files:
            findings.append(
                Finding(
                    check="stale_files",
                    severity="info",
                    title="No stale temp/log files",
                    message="No stale files found under configured temp/log paths.",
                    metadata={"paths": [str(p) for p in self.paths], "stale_file_days": self.stale_file_days},
                    recoverable=True,
                )
            )
            return findings

        sample = [str(p.relative_to(ROOT)) for p in stale_files[:25]]
        total_bytes = sum(p.stat().st_size for p in stale_files if p.exists())
        finding = Finding(
            check="stale_files",
            severity="warn",
            title="Stale temp/log files detected",
            message=f"Found {len(stale_files)} stale files ({_human_size(total_bytes)}).",
            metadata={
                "count": len(stale_files),
                "total_bytes": total_bytes,
                "sample": sample,
                "stale_file_days": self.stale_file_days,
            },
            recoverable=True,
        )

        if mode == "clean" and safe:
            deleted = 0
            for p in stale_files:
                if dry_run:
                    deleted += 1
                    continue
                try:
                    p.unlink(missing_ok=True)
                    deleted += 1
                except Exception:  # noqa: BLE001
                    pass
            finding.cleaned = deleted > 0
            finding.message += f" {'Would remove' if dry_run else 'Removed'} {deleted} file(s)."
            finding.metadata["deleted"] = deleted
            finding.metadata["dry_run"] = dry_run

        findings.append(finding)
        return findings

    def _check_duplicate_migration_prefixes(self) -> list[Finding]:
        findings: list[Finding] = []
        if not self.migrations_dir.exists() or not self.migrations_dir.is_dir():
            return [
                Finding(
                    check="migration_prefixes",
                    severity="info",
                    title="Migration directory missing",
                    message="No migrations directory found; skipping duplicate prefix scan.",
                    metadata={"path": str(self.migrations_dir)},
                )
            ]

        buckets: dict[str, list[str]] = {}
        for p in self.migrations_dir.glob("*.sql"):
            parts = p.name.split("_", 1)
            if not parts or not parts[0].isdigit():
                continue
            buckets.setdefault(parts[0], []).append(p.name)

        duplicates = {k: v for k, v in buckets.items() if len(v) > 1}
        if duplicates:
            findings.append(
                Finding(
                    check="migration_prefixes",
                    severity="critical",
                    title="Duplicate migration prefixes detected",
                    message=f"Detected {len(duplicates)} duplicate migration prefix group(s).",
                    metadata={"duplicates": duplicates},
                )
            )
        else:
            findings.append(
                Finding(
                    check="migration_prefixes",
                    severity="info",
                    title="Migration prefixes are unique",
                    message="No duplicate numeric migration prefixes found.",
                    metadata={"scanned_files": sum(len(v) for v in buckets.values())},
                )
            )
        return findings

    def _check_oversized_session_logs(self, mode: str, safe: bool, dry_run: bool) -> list[Finding]:
        findings: list[Finding] = []

        def is_session_log(p: Path) -> bool:
            n = p.name.lower()
            if not n.endswith(".log"):
                return False
            return "session" in n or "subagent" in n or "agent" in n

        candidates = [p for p in self._all_candidate_files() if is_session_log(p)]
        oversized = [p for p in candidates if p.exists() and p.stat().st_size > self.oversized_log_bytes]

        if not oversized:
            findings.append(
                Finding(
                    check="oversized_session_logs",
                    severity="info",
                    title="No oversized session logs",
                    message="No session log exceeded configured size threshold.",
                    metadata={
                        "threshold_bytes": self.oversized_log_bytes,
                        "threshold_mb": round(self.oversized_log_bytes / (1024 * 1024), 2),
                        "candidates": len(candidates),
                    },
                    recoverable=True,
                )
            )
            return findings

        max_bytes = max(p.stat().st_size for p in oversized)
        severity = "critical" if max_bytes > (self.oversized_log_bytes * 2) else "warn"
        finding = Finding(
            check="oversized_session_logs",
            severity=severity,
            title="Oversized session logs detected",
            message=f"Found {len(oversized)} oversized session logs.",
            metadata={
                "threshold_bytes": self.oversized_log_bytes,
                "logs": [
                    {
                        "path": str(p.relative_to(ROOT)),
                        "bytes": p.stat().st_size,
                    }
                    for p in oversized[:50]
                ],
            },
            recoverable=True,
        )

        if mode == "clean" and safe:
            processed = 0
            for p in oversized:
                if dry_run:
                    processed += 1
                    continue
                try:
                    # Truncate in-place to preserve log handles.
                    with p.open("w", encoding="utf-8"):
                        pass
                    processed += 1
                except Exception:  # noqa: BLE001
                    pass
            finding.cleaned = processed > 0
            finding.message += f" {'Would truncate' if dry_run else 'Truncated'} {processed} log file(s)."
            finding.metadata["processed"] = processed
            finding.metadata["dry_run"] = dry_run

        findings.append(finding)
        return findings

    def log_event(self, summary: SweepSummary) -> None:
        payload = {
            "mode": summary.mode,
            "safe": summary.safe,
            "dry_run": summary.dry_run,
            "risk_score": summary.risk_score,
            "counts": summary.counts,
            "cleaned_items": summary.cleaned_items,
            "findings": [asdict(f) for f in summary.findings],
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }

        metadata = json.dumps(payload, separators=(",", ":")).replace("'", "''")
        message = f"System hygiene sweep {summary.mode}: risk={summary.risk_score}".replace("'", "''")

        sql = f"""
        INSERT INTO cortana_events (event_type, source, severity, message, metadata)
        VALUES (
            'system_hygiene',
            'tools/hygiene/sweep.py',
            '{self._overall_severity(summary)}',
            '{message}',
            '{metadata}'::jsonb
        );
        """

        env = os.environ.copy()
        env["PATH"] = f"/opt/homebrew/opt/postgresql@17/bin:{env.get('PATH', '')}"
        proc = subprocess.run(["psql", self.db, "-v", "ON_ERROR_STOP=1", "-c", sql], text=True, capture_output=True, env=env)
        if proc.returncode != 0 and self.verbose:
            print(f"[warn] failed to write cortana_events: {proc.stderr.strip() or proc.stdout.strip()}")

    @staticmethod
    def _overall_severity(summary: SweepSummary) -> str:
        if summary.counts.get("critical", 0) > 0:
            return "critical"
        if summary.counts.get("warn", 0) > 0:
            return "warning"
        return "info"


def _human_size(size: int) -> str:
    units = ["B", "KB", "MB", "GB", "TB"]
    val = float(size)
    idx = 0
    while val >= 1024 and idx < len(units) - 1:
        val /= 1024
        idx += 1
    if idx == 0:
        return f"{int(val)} {units[idx]}"
    return f"{val:.1f} {units[idx]}"


def _print_human(summary: SweepSummary) -> None:
    print(f"mode={summary.mode} safe={summary.safe} dry_run={summary.dry_run}")
    print(
        "risk_score={} info={} warn={} critical={} cleaned={}".format(
            summary.risk_score,
            summary.counts.get("info", 0),
            summary.counts.get("warn", 0),
            summary.counts.get("critical", 0),
            summary.cleaned_items,
        )
    )
    for f in summary.findings:
        cleaned = " cleaned" if f.cleaned else ""
        print(f"- [{f.severity}] {f.check}: {f.title}{cleaned}")
        print(f"  {f.message}")


def _summary_to_json(summary: SweepSummary) -> str:
    payload = {
        "mode": summary.mode,
        "safe": summary.safe,
        "dry_run": summary.dry_run,
        "risk_score": summary.risk_score,
        "counts": summary.counts,
        "cleaned_items": summary.cleaned_items,
        "findings": [asdict(f) for f in summary.findings],
    }
    return json.dumps(payload, indent=2, sort_keys=True)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="System hygiene sweep")
    sub = parser.add_subparsers(dest="command")

    audit = sub.add_parser("audit", help="Detect hygiene issues only")
    clean = sub.add_parser("clean", help="Cleanup recoverable low-risk items")
    report = sub.add_parser("report", help="Generate hygiene report")

    for p in (parser, audit, clean, report):
        p.add_argument("--stale-session-minutes", type=int, default=180)
        p.add_argument("--stale-file-days", type=int, default=7)
        p.add_argument("--oversized-log-mb", type=int, default=25)
        p.add_argument("--migrations-dir", default=str(DEFAULT_MIGRATIONS_DIR))
        p.add_argument("--db", default="cortana")
        p.add_argument(
            "--paths",
            nargs="*",
            default=[str(p) for p in DEFAULT_PATHS],
            help="Temp/log paths to scan",
        )
        p.add_argument("--no-log-event", action="store_true", help="Skip cortana_events insert")
        p.add_argument("--verbose", action="store_true")

    clean.add_argument("--safe", action="store_true", help="Required safety gate for cleanup")
    clean.add_argument("--dry-run", action="store_true", help="Show cleanup actions without changing files")
    report.add_argument("--json", action="store_true", help="Emit machine-readable JSON")

    args = parser.parse_args()
    if args.command is None:
        args.command = "audit"
    return args


def main() -> int:
    args = parse_args()

    if args.command == "clean" and not args.safe:
        print("error: clean requires --safe")
        return 2

    paths = [Path(p).expanduser().resolve() for p in args.paths]
    sweep = HygieneSweep(
        stale_session_minutes=args.stale_session_minutes,
        stale_file_days=args.stale_file_days,
        oversized_log_mb=args.oversized_log_mb,
        paths=paths,
        migrations_dir=Path(args.migrations_dir).expanduser().resolve(),
        db=args.db,
        verbose=args.verbose,
    )

    summary = sweep.run(
        mode=args.command,
        safe=bool(getattr(args, "safe", False)),
        dry_run=bool(getattr(args, "dry_run", False)),
    )

    if args.command == "report" and args.json:
        print(_summary_to_json(summary))
    else:
        _print_human(summary)

    if not args.no_log_event:
        sweep.log_event(summary)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/hygiene/sweep.ts
#!/usr/bin/env npx tsx

import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { runPsql } from "../lib/db.js";

type Severity = "info" | "warn" | "critical";

type Finding = {
  check: string;
  severity: Severity;
  title: string;
  message: string;
  metadata: Record<string, unknown>;
  recoverable?: boolean;
  cleaned?: boolean;
};

type Args = {
  command: "audit" | "clean" | "report";
  safe: boolean;
  dryRun: boolean;
  json: boolean;
  noLogEvent: boolean;
  verbose: boolean;
  staleSessionMinutes: number;
  staleFileDays: number;
  oversizedLogMb: number;
  migrationsDir: string;
  db: string;
  paths: string[];
};

const ROOT = path.resolve(path.dirname(new URL(import.meta.url).pathname), "..", "..");
const DEFAULT_PATHS = [path.join(ROOT, "tmp"), path.join(ROOT, "logs"), path.join(ROOT, "cortical-loop", "logs")];
const DEFAULT_MIGRATIONS_DIR = path.join(ROOT, "migrations");
const SEVERITY_WEIGHT: Record<Severity, number> = { info: 1, warn: 4, critical: 10 };

function parseArgs(argv: string[]): Args {
  const args: Args = {
    command: "audit",
    safe: false,
    dryRun: false,
    json: false,
    noLogEvent: false,
    verbose: false,
    staleSessionMinutes: 180,
    staleFileDays: 7,
    oversizedLogMb: 25,
    migrationsDir: DEFAULT_MIGRATIONS_DIR,
    db: "cortana",
    paths: [...DEFAULT_PATHS],
  };

  if (argv[0] === "audit" || argv[0] === "clean" || argv[0] === "report") {
    args.command = argv.shift() as Args["command"];
  }

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--safe") args.safe = true;
    else if (arg === "--dry-run") args.dryRun = true;
    else if (arg === "--json") args.json = true;
    else if (arg === "--no-log-event") args.noLogEvent = true;
    else if (arg === "--verbose") args.verbose = true;
    else if (arg === "--stale-session-minutes") args.staleSessionMinutes = Number(argv[++i] ?? args.staleSessionMinutes);
    else if (arg === "--stale-file-days") args.staleFileDays = Number(argv[++i] ?? args.staleFileDays);
    else if (arg === "--oversized-log-mb") args.oversizedLogMb = Number(argv[++i] ?? args.oversizedLogMb);
    else if (arg === "--migrations-dir") args.migrationsDir = path.resolve(argv[++i] ?? args.migrationsDir);
    else if (arg === "--db") args.db = argv[++i] ?? args.db;
    else if (arg === "--paths") {
      const values: string[] = [];
      while (argv[i + 1] && !argv[i + 1].startsWith("--")) values.push(path.resolve(argv[++i]));
      if (values.length > 0) args.paths = values;
    }
  }

  return args;
}

function readOpenClawSubagents(): Record<string, unknown> | null {
  const proc = spawnSync("openclaw", ["subagents", "list", "--json"], {
    cwd: ROOT,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
  if (proc.status !== 0) return null;
  try {
    const parsed = JSON.parse(proc.stdout || "{}");
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed as Record<string, unknown> : null;
  } catch {
    return null;
  }
}

function checkSubagents(args: Args): Finding[] {
  const data = readOpenClawSubagents();
  if (!data) {
    return [{
      check: "subagent_sessions",
      severity: "info",
      title: "Subagent runtime data unavailable",
      message: "openclaw subagents list --json was unavailable; no queue fallback is used.",
      metadata: { threshold_minutes: args.staleSessionMinutes },
    }];
  }

  const staleMs = args.staleSessionMinutes * 60 * 1000;
  const active = Array.isArray(data.active) ? data.active as Array<Record<string, unknown>> : [];
  const recent = Array.isArray(data.recent) ? data.recent as Array<Record<string, unknown>> : [];
  const findings: Finding[] = [];
  const stale = active.filter((session) => Number(session.runtimeMs ?? 0) > staleMs);

  findings.push(stale.length > 0 ? {
    check: "subagent_sessions",
    severity: "critical",
    title: "Stale active subagent sessions",
    message: `${stale.length} running subagent session(s) exceeded stale threshold.`,
    metadata: { threshold_minutes: args.staleSessionMinutes, sessions: stale.slice(0, 25) },
  } : {
    check: "subagent_sessions",
    severity: "info",
    title: "No stale active subagent sessions",
    message: "All active subagent sessions are within freshness threshold.",
    metadata: { active_count: active.length, threshold_minutes: args.staleSessionMinutes },
  });

  const failed = recent.filter((session) => ["failed", "timeout"].includes(String(session.status ?? "")));
  if (failed.length > 0) {
    findings.push({
      check: "subagent_sessions",
      severity: "warn",
      title: "Recent failed or timed-out subagent sessions",
      message: `Detected ${failed.length} recent subagent failure(s).`,
      metadata: { sessions: failed.slice(0, 25) },
    });
  }

  return findings;
}

function listFiles(basePaths: string[]): string[] {
  const files: string[] = [];
  const walk = (filePath: string) => {
    if (!fs.existsSync(filePath)) return;
    const stat = fs.statSync(filePath);
    if (stat.isFile()) {
      files.push(filePath);
      return;
    }
    if (!stat.isDirectory()) return;
    for (const child of fs.readdirSync(filePath)) walk(path.join(filePath, child));
  };
  for (const base of basePaths) walk(base);
  return files;
}

function checkStaleFiles(args: Args): Finding {
  const now = Date.now();
  const staleMs = args.staleFileDays * 24 * 60 * 60 * 1000;
  const stale = listFiles(args.paths).filter((filePath) => now - fs.statSync(filePath).mtimeMs >= staleMs);

  if (stale.length === 0) {
    return {
      check: "stale_files",
      severity: "info",
      title: "No stale temp or log files",
      message: "No stale files found under configured temp/log paths.",
      metadata: { paths: args.paths, stale_file_days: args.staleFileDays },
      recoverable: true,
    };
  }

  let deleted = 0;
  if (args.command === "clean" && args.safe) {
    for (const filePath of stale) {
      if (!args.dryRun) fs.rmSync(filePath, { force: true });
      deleted += 1;
    }
  }

  return {
    check: "stale_files",
    severity: "warn",
    title: "Stale temp or log files detected",
    message: `Found ${stale.length} stale file(s).${deleted ? ` ${args.dryRun ? "Would remove" : "Removed"} ${deleted}.` : ""}`,
    metadata: { count: stale.length, sample: stale.slice(0, 25).map((filePath) => path.relative(ROOT, filePath)), deleted, dry_run: args.dryRun },
    recoverable: true,
    cleaned: deleted > 0,
  };
}

function checkMigrationPrefixes(args: Args): Finding {
  if (!fs.existsSync(args.migrationsDir)) {
    return {
      check: "migration_prefixes",
      severity: "info",
      title: "Migration directory missing",
      message: "No migrations directory found; skipping duplicate prefix scan.",
      metadata: { path: args.migrationsDir },
    };
  }

  const buckets = new Map<string, string[]>();
  for (const name of fs.readdirSync(args.migrationsDir)) {
    if (!name.endsWith(".sql")) continue;
    const prefix = name.split("_", 1)[0];
    if (!/^\d+$/.test(prefix)) continue;
    buckets.set(prefix, [...(buckets.get(prefix) ?? []), name]);
  }
  const duplicates = [...buckets.entries()].filter(([, names]) => names.length > 1);

  return duplicates.length > 0 ? {
    check: "migration_prefixes",
    severity: "critical",
    title: "Duplicate migration prefixes detected",
    message: `Detected ${duplicates.length} duplicate migration prefix group(s).`,
    metadata: { duplicates: Object.fromEntries(duplicates) },
  } : {
    check: "migration_prefixes",
    severity: "info",
    title: "Migration prefixes are unique",
    message: "No duplicate numeric migration prefixes found.",
    metadata: { scanned_files: [...buckets.values()].flat().length },
  };
}

function checkOversizedLogs(args: Args): Finding {
  const maxBytes = args.oversizedLogMb * 1024 * 1024;
  const logs = listFiles(args.paths).filter((filePath) => {
    const name = path.basename(filePath).toLowerCase();
    return name.endsWith(".log") && (name.includes("session") || name.includes("subagent") || name.includes("agent"));
  });
  const oversized = logs.filter((filePath) => fs.statSync(filePath).size > maxBytes);

  if (oversized.length === 0) {
    return {
      check: "oversized_session_logs",
      severity: "info",
      title: "No oversized session logs",
      message: "No session log exceeded configured size threshold.",
      metadata: { threshold_mb: args.oversizedLogMb, candidates: logs.length },
      recoverable: true,
    };
  }

  let processed = 0;
  if (args.command === "clean" && args.safe) {
    for (const filePath of oversized) {
      if (!args.dryRun) fs.writeFileSync(filePath, "");
      processed += 1;
    }
  }

  return {
    check: "oversized_session_logs",
    severity: "warn",
    title: "Oversized session logs detected",
    message: `Found ${oversized.length} oversized session log(s).${processed ? ` ${args.dryRun ? "Would truncate" : "Truncated"} ${processed}.` : ""}`,
    metadata: { threshold_mb: args.oversizedLogMb, logs: oversized.slice(0, 25).map((filePath) => path.relative(ROOT, filePath)), processed, dry_run: args.dryRun },
    recoverable: true,
    cleaned: processed > 0,
  };
}

function summarize(args: Args, findings: Finding[]) {
  const counts: Record<Severity, number> = { info: 0, warn: 0, critical: 0 };
  let weighted = 0;
  let cleanedItems = 0;
  for (const finding of findings) {
    counts[finding.severity] += 1;
    weighted += SEVERITY_WEIGHT[finding.severity];
    if (finding.cleaned) cleanedItems += 1;
  }
  const riskScore = findings.length === 0 ? 0 : Math.min(100, Math.round((weighted / (findings.length * SEVERITY_WEIGHT.critical)) * 100));
  return { mode: args.command, safe: args.safe, dry_run: args.dryRun, risk_score: riskScore, counts, cleaned_items: cleanedItems, findings };
}

function logEvent(args: Args, summary: ReturnType<typeof summarize>): void {
  const severity = summary.counts.critical > 0 ? "critical" : summary.counts.warn > 0 ? "warning" : "info";
  const metadata = JSON.stringify({ ...summary, timestamp: new Date().toISOString() }).replace(/'/g, "''");
  const message = `System hygiene sweep ${summary.mode}: risk=${summary.risk_score}`.replace(/'/g, "''");
  const proc = runPsql(`
INSERT INTO cortana_events (event_type, source, severity, message, metadata)
VALUES ('system_hygiene', 'tools/hygiene/sweep.ts', '${severity}', '${message}', '${metadata}'::jsonb);
`, { db: args.db });
  if (proc.status !== 0 && args.verbose) {
    console.error(`[warn] failed to write cortana_events: ${(proc.stderr || proc.stdout || "").trim()}`);
  }
}

function printHuman(summary: ReturnType<typeof summarize>): void {
  console.log(`mode=${summary.mode} safe=${summary.safe} dry_run=${summary.dry_run}`);
  console.log(`risk_score=${summary.risk_score} info=${summary.counts.info} warn=${summary.counts.warn} critical=${summary.counts.critical} cleaned=${summary.cleaned_items}`);
  for (const finding of summary.findings) {
    console.log(`- [${finding.severity}] ${finding.check}: ${finding.title}`);
    console.log(`  ${finding.message}`);
  }
}

function main(): number {
  const args = parseArgs(process.argv.slice(2));
  if (args.command === "clean" && !args.safe) {
    console.error("error: clean requires --safe");
    return 2;
  }

  const findings = [
    ...checkSubagents(args),
    checkStaleFiles(args),
    checkMigrationPrefixes(args),
    checkOversizedLogs(args),
  ];
  const summary = summarize(args, findings);

  if (args.command === "report" && args.json) console.log(JSON.stringify(summary, null, 2));
  else printHuman(summary);

  if (!args.noLogEvent) logEvent(args, summary);
  return 0;
}

process.exit(main());
